demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c

```diff
@@ -0,0 +1,1354 @@
+                              MQTTPublishInfo_t * pPublishInfo )
+{
+    assert( pPublishInfo != NULL );
+    assert( pContext != NULL );
+
+    OtaEventData_t * pData;
+    OtaEventMsg_t eventMsg = { 0 };
+
+    LogInfo( ( "Received data message callback, size %zu.\n\n", pPublishInfo->payloadLength ) );
+
+    pData = otaEventBufferGet();
+
+    if( pData != NULL )
+    {
+        memcpy( pData->data, pPublishInfo->pPayload, pPublishInfo->payloadLength );
+        pData->dataLength = pPublishInfo->payloadLength;
+        eventMsg.eventId = OtaAgentEventReceivedFileBlock;
+        eventMsg.pEventData = pData;
+
+        /* Send job document received event. */
+        OTA_SignalEvent( &eventMsg );
+    }
+    else
+    {
+        LogError( ( "Error: No OTA data buffers available.\r\n" ) );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static void mqttEventCallback( MQTTContext_t * pMqttContext,
+                               MQTTPacketInfo_t * pPacketInfo,
+                               MQTTDeserializedInfo_t * pDeserializedInfo )
+{
+    assert( pMqttContext != NULL );
+    assert( pPacketInfo != NULL );
+    assert( pDeserializedInfo != NULL );
+
+    /* Handle incoming publish. The lower 4 bits of the publish packet
+     * type is used for the dup, QoS, and retain flags. Hence masking
+     * out the lower bits to check if the packet is publish. */
+    if( ( pPacketInfo->type & 0xF0U ) == MQTT_PACKET_TYPE_PUBLISH )
+    {
+        assert( pDeserializedInfo->pPublishInfo != NULL );
+        /* Handle incoming publish. */
+        SubscriptionManager_DispatchHandler( pMqttContext, pDeserializedInfo->pPublishInfo );
+    }
+    else
+    {
+        /* Handle other packets. */
+        switch( pPacketInfo->type )
+        {
+            case MQTT_PACKET_TYPE_SUBACK:
+                LogInfo( ( "Received SUBACK.\n\n" ) );
+                break;
+
+            case MQTT_PACKET_TYPE_UNSUBACK:
+                LogInfo( ( "Received UNSUBACK.\n\n" ) );
+                break;
+
+            case MQTT_PACKET_TYPE_PINGRESP:
+
+                /* Nothing to be done from application as library handles
+                 * PINGRESP. */
+                LogWarn( ( "PINGRESP should not be handled by the application "
+                           "callback when using MQTT_ProcessLoop.\n\n" ) );
+                break;
+
+            case MQTT_PACKET_TYPE_PUBACK:
+                LogInfo( ( "PUBACK received for packet id %u.\n\n",
+                           pDeserializedInfo->packetIdentifier ) );
+                break;
+
+            /* Any other packet type is invalid. */
+            default:
+                LogError( ( "Unknown packet type received:(%02x).\n\n",
+                            pPacketInfo->type ) );
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static int32_t generateRandomNumber()
+{
+    return( rand() % ( INT32_MAX ) );
+}
```

Comment:
The `backoffAlgorithm` library requires a `uint32_t` value for the random number
```suggestion
static uint32_t generateRandomNumber()
{
    return( rand() );
}
```